node/src/components/storage.rs

```diff
@@ -1196,19 +1219,67 @@ impl StorageInner {
         Ok(outcome)
     }
 
+    /// Puts block and its deploys into storage.
+    ///
+    /// Returns `Ok` only if the block and all deploys were successfully written.
+    pub fn put_block_and_deploys(
+        &self,
+        block_and_deploys: &BlockAndDeploys,
+    ) -> Result<(), FatalStorageError> {
+        let BlockAndDeploys { block, deploys } = block_and_deploys;
+
+        block.verify(self.verifiable_chunked_hash_activation)?;
+        let mut txn = self.env.begin_rw_txn()?;
+        match self.write_validated_block(&mut txn, block) {
+            Ok(true) => {}
+            Ok(false) => {
+                txn.abort();
+                return Err(FatalStorageError::FailedToOverwriteBlock);
+            }
+            Err(error) => {
+                txn.abort();
+                return Err(error);
+            }
+        }
+
+        for deploy in deploys {
+            let _ = txn.put_value(self.deploy_db, deploy.id(), deploy, false)?;
+        }
+        txn.commit()?;
+
+        Ok(())
+    }
+
     /// Retrieves a block by hash.
     pub fn read_block(&self, block_hash: &BlockHash) -> Result<Option<Block>, FatalStorageError> {
         self.get_single_block(&mut self.env.begin_ro_txn()?, block_hash)
     }
 
-    /// Writes a block to storage, updating indices as necessary
+    /// Writes a block to storage, updating indices as necessary.
+    ///
     /// Returns `Ok(true)` if the block has been successfully written, `Ok(false)` if a part of it
     /// couldn't be written because it already existed, and `Err(_)` if there was an error.
     pub fn write_block(&self, block: &Block) -> Result<bool, FatalStorageError> {
         // Validate the block prior to inserting it into the database
         block.verify(self.verifiable_chunked_hash_activation)?;
         let mut txn = self.env.begin_rw_txn()?;
-        // Write the block body
+        let result = self.write_validated_block(&mut txn, block);
+        match &result {
+            Ok(false) | Err(_) => txn.abort(),
+            Ok(true) => txn.commit()?,
+        }
+        result
```

Comment:
nit: could be
```
Ok(self
    .read_deploys(deploys_count, deploy_hashes)?
    .map(|deploys| BlockAndDeploys { block, deploys }))
```
or
```
self.read_deploys(deploys_count, deploy_hashes)
    .map(|maybe_deploys| {
        maybe_deploys.map(|deploys| BlockAndDeploys { block, deploys })
    })
```

Reply:
Fixed in [ecf4e89](https://github.com/casper-network/casper-node/pull/3010/commits/ecf4e895ee12a1b1c67900fe7d519bc4ed9dc658).